L1Trigger/TrackerTFP/src/DataFormats.cc

```diff
@@ -0,0 +1,817 @@
+#include "L1Trigger/TrackerTFP/interface/DataFormats.h"
+#include "L1Trigger/TrackTrigger/interface/StubPtConsistency.h"
+
+#include <vector>
+#include <deque>
+#include <cmath>
+#include <tuple>
+#include <iterator>
+#include <algorithm>
+#include <string>
+#include <iostream>
+
+using namespace std;
+using namespace edm;
+using namespace trackerDTC;
+
+namespace trackerTFP {
+
+  // default constructor, trying to needs space as proper constructed object
+  DataFormats::DataFormats() :
+    numDataFormats_(0),
+    formats_(+Variable::end, std::vector<DataFormat*>(+Process::end, nullptr)),
+    numUnusedBitsStubs_(+Process::end, TTBV::S_),
+    numUnusedBitsTracks_(+Process::end, TTBV::S_),
+    numChannel_(+Process::end, 0)
+  {
+    setup_ = nullptr;
+    countFormats();
+    dataFormats_.reserve(numDataFormats_);
+    numStreams_.reserve(+Process::end);
+    numStreamsStubs_.reserve(+Process::end);
+    numStreamsTracks_.reserve(+Process::end);
+  }
+
+  // method to count number of unique data formats
+  template<Variable v = Variable::begin, Process p = Process::begin>
+  void DataFormats::countFormats() {
+    if constexpr(config_[+v][+p] == p)
+      numDataFormats_++;
+    if constexpr(++p != Process::end)
+      countFormats<v, ++p>();
+    else if constexpr(++v != Variable::end)
+      countFormats<++v>();
+  }
+
+  // proper constructor
+  DataFormats::DataFormats(const ParameterSet& iConfig, const Setup* setup) : DataFormats() {
+    iConfig_ = iConfig;
+    setup_ = setup;
+    fillDataFormats();
+    for (const Process p : Processes)
+      for (const Variable v : stubs_[+p])
+        numUnusedBitsStubs_[+p] -= formats_[+v][+p] ? formats_[+v][+p]->width() : 0;
+    for (const Process p : Processes)
+      for (const Variable v : tracks_[+p])
+        numUnusedBitsTracks_[+p] -= formats_[+v][+p] ? formats_[+v][+p]->width() : 0;
+    numChannel_[+Process::dtc] = setup_->numDTCsPerRegion();
+    numChannel_[+Process::pp] = setup_->numDTCsPerTFP();
+    numChannel_[+Process::gp] = setup_->numSectors();
+    numChannel_[+Process::ht] = setup_->htNumBinsInv2R();
+    numChannel_[+Process::mht] = setup_->htNumBinsInv2R();
+    numChannel_[+Process::sf] = setup_->htNumBinsInv2R();
+    numChannel_[+Process::kfin] = setup_->kfNumWorker() * setup_->numLayers();
```

Comment:
Isn't the number of channels for KFin proportional to "numLayers + 1", since one channel is used for the track?